frontends/p4/def_use.cpp

```diff
@@ -759,6 +763,39 @@ void ComputeWriteSet::visitVirtualMethods(const IR::IndexedVector<IR::Declaratio
     }
 }
 
+std::size_t P4::loc_t::hash() const {
+    if (!parent) return Util::Hash{}(node->id);
+
+    return Util::Hash{}(node->id, parent->hash());
```

Comment:
Do we have an idea how "deep" the recursion here could be?

Reply:
It is as deep as the deepest IR::Node in an IR::Node DAG. In many of the P4-16 programs that I have analyzed it does not get egregiously deep.

Reply:
Well... this is quite a lot actually and it could be very deep especially after inlining. I would probably check the effects of memoization. We can use zero hash value as a thombstone, in a very rare case of zero value hash collision, well, we'd have a single level of recursion.

Reply:
I tried unsuccessfully to memoize the hashes: https://github.com/kfcripps/p4c/commit/04b7a4e7b4aba8d5ba3bf40c010db15dcb55cebf

Let me know if you have any suggestions on how to accomplish this

Reply:
Your suggestion worked. Thanks!